### ISSUE TITLE
1D GANs in Gluon

### ISSUE THREAD

**Issue by SumNeuron** [apache/incubator-mxnet#8375]
#Description 
The tutorial for GANs in MXNet can be found here: 

https://mxnet.incubator.apache.org/tutorials/unsupervised_learning/gan.html

The example uses the symbol and module modules, which to my understanding have been replaced(?) by the gluon module.

So lets say I wanted to make the most simple gans utilizing only gluon. We will use fake data (lets say a 1d array)

#Code 

```
import mxnet as mx
from mxnet.gluon import nn
import mxnet.ndarray as nd
import numpy as np


batch_size = 100
epochs = 1
ctx = mx.cpu()


# super simple generator
gen = nn.Sequential()
gen.add(nn.Dense(50, activation='relu'))
gen.add(nn.Dense(25, activation='relu'))
gen.add(nn.Dense(10, activation='relu'))

# super simple discriminator 
dis = nn.Sequential()
dis.add(nn.Dense(200, activation='relu'))
dis.add(nn.Dense(100, activation='relu'))
dis.add(nn.Dense(50, activation='relu'))
dis.add(nn.Dense(25, activation='relu'))
dis.add(nn.Dense(2, activation='relu'))

# simulated "real" data
real_samples = 1000
real_data_length = 10
real_mean = 0
real_standard_deviation = 1
real_data = nd.random_normal(loc=real_mean, scale=real_standard_deviation, shape=(real_samples, real_data_length))



# Iterators...
class RandIter(mx.io.DataIter):
    def __init__(self, batch_size, ndim):
        self.batch_size = batch_size
        self.ndim = ndim
        self.provide_data = [('rand', (batch_size, ndim))]
        self.provide_label = []
        
    def iter_next(self):
        return True
    
    def getdata(self):
        return [nd.random_normal(0, 1, shape=(self.batch_size, self.ndim))]

# Produce 'simulated' data
Z = 100
rand_iter = RandIter(batch_size, Z)



# Hyperperameters
sigma = 0.02
lr = 0.0002
beta1 = 0.5



# Train Data 
print('Training...')
for epoch in range(epochs):
    # what goes here?

```

**Comment by szha** [comment]
https://github.com/zackchase/mxnet-the-straight-dope/tree/master/chapter14_generative-adversarial-networks

**Comment by SumNeuron** [comment]
Thanks, guess I missed this one.
I have a few questions about the training implementation. 

I have put them as inline comments from the copy-pasted code, which start as #!


```
for epoch in range(10):
    tic = time.time()
    train_data.reset()
    for i, batch in enumerate(train_data):
        ############################
        # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
        ###########################
        # train with real_t
        data = batch.data[0].as_in_context(ctx)
        noise = nd.random_normal(shape=(batch_size, 2), ctx=ctx) #! this replaces the purpose of Z in dcgans?

        with autograd.record():
            real_output = netD(data)
            errD_real = loss(real_output, real_label)
            
            fake = netG(noise)
            fake_output = netD(fake.detach()) #! why is this detached when in the dcgan it is not?
            errD_fake = loss(fake_output, fake_label)
            errD = errD_real + errD_fake
            errD.backward() 

        trainerD.step(batch_size)
        metric.update([real_label,], [real_output,]) #! why are these included in the record statement in dcgan?
        metric.update([fake_label,], [fake_output,])

        ############################
        # (2) Update G network: maximize log(D(G(z)))
        ###########################
        with autograd.record():
            output = netD(fake)
            errG = loss(output, real_label) 
            errG.backward()

        trainerG.step(batch_size)
```

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by indhub** [comment]
This is not an issue. Please use the [user forum](https://discuss.mxnet.io/) for how-to questions. Thanks!
